File: 프로그래머스/월간 코드 챌린지 시즌1/쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
v = []

def merge(start_i, start_j, N):
    global v
    flag = True
    a = v[start_i][start_j]
    for i in range(N):
        for j in range(N):
            if v[start_i+i][start_j+j] != a:
                flag = False
                break
        if not flag: break

    if flag:
        if a == 1: return [0, 1]
        else: return [1, 0]
    else:
        answer = [0, 0]

        tmp = merge(start_i, start_j, N//2)
        answer[0] += tmp[0]
        answer[1] += tmp[1]

        tmp = merge(start_i, start_j+N//2, N//2)
        answer[0] += tmp[0]
        answer[1] += tmp[1]

        tmp = merge(start_i+N//2, start_j, N//2)
        answer[0] += tmp[0]
        answer[1] += tmp[1]

        tmp = merge(start_i+N//2, start_j+N//2, N//2)
        answer[0] += tmp[0]
        answer[1] += tmp[1]

        return answer
    
def solution(arr):
    global v
    v = arr
    answer = [0, 0]

    p = merge(0, 0, len(arr))
    return [p[0], p[1]]

File: 프로그래머스/월간 코드 챌린지 시즌1/쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> v;

pair<int,int> merge(int start_i, int start_j, int N){
    bool flag = true;
    int a = v[start_i][start_j];
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (v[start_i + i][start_j + j] != a) {
                flag = false;
                break;
            }
        }
        if (!flag) break;
    }
    
    // NxN 크기가 다 같은 숫자라면
    if(flag){
        if(a == 1) return {0, 1};
        else return {1, 0};
    }
    // 아니면 네 방향으로 쪼개기
    else{
        pair<int,int> answer = {0, 0};
        
        // 왼쪽 위 방향
        pair<int,int> tmp = merge(start_i, start_j, N/2);
        answer.first += tmp.first;
        answer.second += tmp.second;
        
        // 오른쪽 위 방향
        tmp = merge(start_i, start_j+N/2, N/2);
        answer.first += tmp.first;
        answer.second += tmp.second;
        
        // 왼쪽 아래 방향
        tmp = merge(start_i+N/2, start_j, N/2);
        answer.first += tmp.first;
        answer.second += tmp.second;
        
        // 오른쪽 아래 방향
        tmp = merge(start_i+N/2, start_j+N/2, N/2);
        answer.first += tmp.first;
        answer.second += tmp.second;
        
        return answer;
    }
}

vector<int> solution(vector<vector<int>> arr) {
    int N = arr.size(); // 정사각형 크기
    v = arr;
    pair<int,int> answer = {0, 0};  // <0의 개수, 1의 개수>
    
    pair<int,int> p = merge(0, 0, N);
    return {p.first, p.second};
}
